### PR TITLE
Rewrite framed replay input handler for robustness

### LINKED ISSUE #6150: FramedReplayHandler doesn't respect direction for the current and next frame
From the middle of two frames - 0 -> 1, during rewind the "current" frame remains 0 and the "next" frame remains 1, although they should be reversed.

There's a few other edge cases to consider, such as rewinding from before the first frame, playing back normally from before the first frame, and vice versa for the last frame.

I've already done work on this in https://github.com/ppy/osu/compare/master...smoogipoo:fix-framedreplayhandler which should already be mostly working, but the test cases need a lot of work.

### LINKED ISSUE #10811: Rewinding to beginning of autoplay/replay will miss first hit object
**Describe the bug:** In any game mode, rewinding to beginning or rewinding at lead-in time will go straight to the first object and will miss. I think it should rewind to the start of the lead-in time to prevent misses and weird cut off in replays at the start.

**Screenshots or videos showing encountered issue:** 
3AtAvna1ff

**osu!lazer version:** 2020.1109.0

**Logs:** not needed

### PR DESCRIPTION
**Pull request by ekrctb**
### Motivation

- The previous code was full of nullable types and it was cumbersome to use (by the derived classes).
- There was no explicit invariant and it was difficult to reason about.
- Fixes #6150. Fixes #10811.

### Methods

I tried to make less breaking changes as possible, but there are some changes.
The getters of the frames now throw when the replay is empty, instead of returning `null`.
The values of `CurrentFrame` and `NextFrame` are changed. Namely, the ordering of `NextFrame.Time` and `CurrentFrame.Time` was dependent on the current direction. Now, it should always satisfy `CurrentFrame.Time <= CurrentTime <= NextFrame.Time` except at the start/end.

### Result

- The issue #6150 can be visually confirmed by repeatedly rewinding a catch replay and noticing the catcher goes out-of-bounds in master but that is fixed with this PR.
- Luckily, the change of `CurrentFrame` and `NextFrame` doesn't break existing deriving classes if the template code pattern usage of interpolation is used.
  - This is due to the symmetry: `interpolate(time, start, end) = interpolate(time, end, start)`.
- The deriving class code can be simplified due to the elimination of nullable types. This is <https://github.com/ppy/osu/commit/bfb7805838be2e777b27919a127040597b7092c0> but not included in this PR because a breaking change is required to simplify more and I would like to have a separate PR.
  - Edit: this change is partially reverted, as it was breaking a certain case. It will be included in the next PR.
- Return value of `SetFrameFromTime` (whether to return null) is also changed, but it only affected the testing code, and also it follows the documentation (returning `null` when the frame is not valid) more correctly.

### Discussion

I wanted to make `ReplayInputHandler` is not called when the replay is empty, using `IsActive` property of `InputHandler`. However, `IsActive` is unimplemented in the osu-framework (dead code) so I had to modify `RulesetInputManager`. Probably it should be supported in `osu-framework`, or `IsActive` shouldn't be used.
`StreamingFramedReplayInputHandlerTest` is removed, as it was duplicate. I added a test to `FramedReplayInputHandlerTest` as a replacement.

### CODE REVIEW

**Comment by peppy** [comment]
Curious if this happens to resolve https://github.com/ppy/osu/issues/10811 as well.

File: osu.Game/Rulesets/UI/RulesetInputManager.cs [hunk 1]
```diff
@@ -100,6 +102,14 @@ public ReplayInputHandler ReplayInputHandler
 
         #endregion
 
+        protected override List<IInput> GetPendingInputs()
+        {
+            if (replayInputHandler != null && !replayInputHandler.IsActive)
+                return new List<IInput>();
```
Comment (peppy) [inline osu.Game/Rulesets/UI/RulesetInputManager.cs:108]: The type of this method is unfortunate in that we can't use a static representation of the empty list. To avoid allocation of the `new List<>` each call, a better option may be to call the base method and then perform a `.Clear()` on the list.

File: osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs [hunk 2]
```diff
@@ -17,65 +16,80 @@ namespace osu.Game.Rulesets.Replays
...
+        /// <summary>
+        /// Whether we have at least one replay frame.
+        /// </summary>
+        public bool HasFrames => Frames.Count != 0;
 
-        protected List<ReplayFrame> Frames => replay.Frames;
+        /// <summary>
+        /// Whether we are waiting for new frames to be received.
+        /// </summary>
+        public bool WaitingNextFrame => !replay.HasReceivedAllFrames && currentFrameIndex == Frames.Count - 1;
```
Comment (peppy) [inline osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:27]:
```suggestion
        public bool WaitingForFrame => !replay.HasReceivedAllFrames && currentFrameIndex == Frames.Count - 1;
```

or `WaitingForNextFrame`.

File: osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs [hunk 3]
```diff
@@ -17,65 +16,80 @@ namespace osu.Game.Rulesets.Replays
...
+                    throw new InvalidOperationException($"Attempted to get {nameof(CurrentFrame)} of an empty replay");
 
-                return (TFrame)Frames[currentFrameIndex.Value];
+                return (TFrame)Frames[Math.Max(0, currentFrameIndex)];
             }
         }
 
+        /// <summary>
+        /// The next frame of the replay.
+        /// The start time is always greater or equals to the start time of <see cref="CurrentFrame"/> regardless of the seeking direction.
```
Comment (peppy) [inline osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:47]:
```suggestion
        /// The start time is always greater or equal to the start time of <see cref="CurrentFrame"/> regardless of the seeking direction.
```

File: osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs [hunk 4]
```diff
@@ -17,65 +16,80 @@ namespace osu.Game.Rulesets.Replays
...
-        public bool FrameAccuratePlayback;
+        private int currentFrameIndex;
 
-        public bool HasFrames => Frames.Count > 0;
+        private const double sixty_frame_time = 1000.0 / 60;
+
+        protected FramedReplayInputHandler(Replay replay)
+        {
+            // This replay frame ordering should be enforced on the Replay type
+            replay.Frames.Sort((x, y) => x.Time.CompareTo(y.Time));
```
Comment (peppy) [inline osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:87]: Is the intention to remove this eventually (in a follow up PR)? Probably best to prefix the comment with `TODO:`.

Also as this is done at construction time, it won't be applied to streaming frames so that's something to consider.

**Review by peppy — CHANGES_REQUESTED** [review]
Just an initial pass. Will do some more reading on the actual logic changes and your questions about the `InputManager.IsActive` usage in a second pass.

**Comment by ekrctb** [comment]
The `TestSceneGameplayRewinding` failure is due to the change of `CurrentFrame` (while rewinding).
`KeyCounter` has `IsCounting` property. This is set to false when the gameplay is in a break and set to true when no longer in the break, but the change is delaying one frame. The ordering of the process is as follows (when rewinding from the end of the map):

1. (New frame)
2. `BreakTracker.Update`: `IsBreakTime = true` because `HasCompleted == true`.
3. `RulesetInputManager` process input actions. It triggers `KeyCounter.OnReleased`, but it is not counted as `IsCounting == false`, which is bound to `IsBreakTime`.
4. `DHO.Update` triggers `RevertResult`, its event listener set `HasCompleted = false`.
5. (New frame)

Previously, the future frame's input actions were used while rewinding, which is semantically wrong (this issue is different from #6150), but the key counter worked because it effectively delayed the input action change to compensate for the one-frame delay above.

Fixing it by changing drawable child ordering seems difficult and fragile.
I think a proper solution is to make `KeyCounter` record past key presses in an explicit list (potentially using `Replay`, but not sure it makes sense) so state rewinding can be done robustly. It would be a separate PR.

**Comment by ekrctb** [comment]
Hopefully I managed to fix the test failure with a little code change. See the commit message for the elaboration.

File: osu.Game/Rulesets/Scoring/JudgementProcessor.cs [hunk 5]
```diff
@@ -134,6 +133,10 @@ void simulate(HitObject obj)
             }
         }
 
-        private void updateHasCompleted() => hasCompleted.Value = JudgedHits == MaxHits;
+        protected override void Update()
+        {
+            base.Update();
+            hasCompleted.Value = JudgedHits == MaxHits && (JudgedHits == 0 || lastAppliedResult.TimeAbsolute < Clock.CurrentTime);
```
Comment (peppy) [inline osu.Game/Rulesets/Scoring/JudgementProcessor.cs:139]: Personally I'm okay with this change, but would like a second opinion from @smoogipoo (see https://github.com/ppy/osu/pull/12376/commits/d0f30b7b422afe385fd1fc0de7a28684bf0ab7ef for reasoning).

**Review by peppy — APPROVED** [review]
I'm really happy with this direction. It reads a lot better and reduces the bar of entry to understanding what is going on with this logic.

The testing changes look correct and match my expectations better than what we had previously.

👍🏻

File: osu.Game/Rulesets/Scoring/JudgementProcessor.cs (hunk 5 above)
Comment (smoogipoo) [reply osu.Game/Rulesets/Scoring/JudgementProcessor.cs:139]: Though it's super dodgy, I think I'm fine with this. I'd like this to not be a thing in the long term though.

**Review by smoogipoo — APPROVED** [review]
